providers/keycloak.go

```diff
@@ -63,37 +68,151 @@ func (p *KeycloakProvider) SetGroup(group string) {
 }
 
-func (p *KeycloakProvider) GetEmailAddress(ctx context.Context, s *sessions.SessionState) (string, error) {
-	json, err := requests.New(p.ValidateURL.String()).
+func (p *KeycloakProvider) SetRoles(roles []string) {
+	p.Roles = roles
+}
+
+func extractRolesFromClaims(claims map[string]interface{}) ([]string, error) {
+	var roleList []string
+
+	if realmRoles, found := claims["realm_access"].(map[string]interface{}); found {
+		if roles, found := realmRoles["roles"]; found {
+			_, ok := roles.([]interface{})
+			if !ok {
+				return nil, errors.New("error parsing realm roles from claims")
+			}
+			for _, r := range roles.([]interface{}) {
+				roleList = append(roleList, fmt.Sprintf("%s", r))
+			}
+		}
+	}
+
+	if clientRoles, found := claims["resource_access"].(map[string]interface{}); found {
+		for name, list := range clientRoles {
+			scopes, ok := list.(map[string]interface{})
+			if !ok {
+				return nil, errors.New("error parsing client roles from claims")
+			}
+			if roles, found := scopes["roles"]; found {
+				for _, r := range roles.([]interface{}) {
+					roleList = append(roleList, fmt.Sprintf("%s:%s", name, r))
+				}
+			}
+		}
+	}
+
+	return roleList, nil
+}
+
+func (p *KeycloakProvider) getUserInfo(ctx context.Context, s *sessions.SessionState) (*simplejson.Json, error) {
+	userInfo, err := requests.New(p.ValidateURL.String()).
 		WithContext(ctx).
 		SetHeader("Authorization", "Bearer "+s.AccessToken).
 		Do().
 		UnmarshalJSON()
 	if err != nil {
 		logger.Errorf("failed making request %s", err)
-		return "", err
+		return nil, fmt.Errorf("error getting user info: %v", err)
 	}
+	return userInfo, nil
+}
 
-	if p.Group != "" {
-		var groups, err = json.Get("groups").Array()
+func getEmailAddress(userInfo *simplejson.Json) (string, error) {
+	var email, err = userInfo.Get("email").String()
+	if err != nil {
+		return "", fmt.Errorf("error getting email address: %v", err)
+	}
+	return email, nil
+}
+
+func getGroups(userInfo *simplejson.Json) ([]string, error) {
+	if g, ok := userInfo.CheckGet("groups"); ok {
+		groups, err := g.StringArray()
 		if err != nil {
-			logger.Printf("groups not found %s", err)
-			return "", err
+			return []string{}, fmt.Errorf("error getting groups: %v", err)
 		}
+		return groups, nil
+	}
+	return []string{}, nil
+}
 
+func getRoles(s *sessions.SessionState) ([]string, error) {
+	claims := make(map[string]interface{})
+	// Decode JWT token without verifying the signature
+	token, err := jwt.ParseSigned(s.AccessToken)
+	if err != nil {
+		return nil, fmt.Errorf("failed to parse token: %v", err)
+	}
+	// Parse claims
+	if err := token.UnsafeClaimsWithoutVerification(&claims); err != nil {
+		logger.Printf("failed to parse claims: %s", err)
+	}
+	roles, err := extractRolesFromClaims(claims)
+	if err != nil {
+		logger.Printf("failed to extract roles: %s", err)
+	}
+	return roles, nil
+}
+
+func (p *KeycloakProvider) isAuthorizedGroupMember(s *sessions.SessionState) (bool, error) {
```

Comment:
This should be unneeded if you set the roles you want like this to take advantage of the global `Authorize` logic. Use a prefixed name like I mentioned above to not collide with groups -- e.g. `role:foorbar`

https://github.com/oauth2-proxy/oauth2-proxy/blob/d67d6e3152e2a1c9614906ee9bd9890f8da9a007/pkg/validation/options.go#L289

Here are the pertinent areas that do this for you:
https://github.com/oauth2-proxy/oauth2-proxy/blob/d67d6e3152e2a1c9614906ee9bd9890f8da9a007/providers/provider_data.go#L56

https://github.com/oauth2-proxy/oauth2-proxy/blob/d67d6e3152e2a1c9614906ee9bd9890f8da9a007/providers/provider_default.go#L102

Reply:
I've done that, I would still appreciate your input on my comment regarding the groups/roles ANY logic on the other thread.